packages/eslint-plugin/src/rules/strict-enums.ts

```diff
@@ -0,0 +1,678 @@
+    },
+    messages: {
+      incorrectIncrement: 'You cannot increment or decrement an enum type.',
+      mismatchedAssignment:
+        'The type of the enum assignment does not match the declared enum type of the variable.',
+      mismatchedComparison:
+        'The two things in the comparison do not have a shared enum type.',
+      mismatchedFunctionArgument:
+        'The {{ ordinal }} argument in the function call does not match the declared enum type of the function signature.',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {},
+        additionalProperties: false,
+      },
+    ],
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const typeChecker = parserServices.program.getTypeChecker();
+
+    // ----------------
+    // Helper functions
+    // ----------------
+
+    /**
+     * If passed an enum member, returns the type of the parent. Otherwise,
+     * returns itself.
+     *
+     * For example:
+     * - `Fruit` --> `Fruit`
+     * - `Fruit.Apple` --> `Fruit`
+     */
+    function getBaseEnumType(type: ts.Type): ts.Type {
+      const symbol = type.getSymbol();
+      if (symbol === undefined) {
+        return type;
+      }
+
+      if (!util.isSymbolFlagSet(symbol, ts.SymbolFlags.EnumMember)) {
+        return type;
+      }
+
+      const { valueDeclaration } = symbol;
+      if (valueDeclaration === undefined) {
+        return type;
+      }
+
+      const parentType = getTypeFromTSNode(valueDeclaration.parent);
+      if (parentType === undefined) {
+        return type;
+      }
+
+      return parentType;
+    }
+
+    /**
+     * A thing can have 0 or more enum types. For example:
+     * - 123 --> []
+     * - {} --> []
+     * - Fruit.Apple --> [Fruit]
+     * - Fruit.Apple | Vegetable.Lettuce --> [Fruit, Vegetable]
+     * - Fruit.Apple | Vegetable.Lettuce | 123 --> [Fruit, Vegetable]
+     * - T extends Fruit --> [Fruit]
+     */
+    function getEnumTypes(type: ts.Type): Set<ts.Type> {
+      /**
+       * First, we get all the parts of the union. For non-union types, this
+       * will be an array with the type in it. For example:
+       * - Fruit --> [Fruit]
+       * - Fruit | Vegetable --> [Fruit, Vegetable]
+       */
+      const subTypes = tsutils.unionTypeParts(type);
+
+      /**
+       * Next, we must resolve generic types with constraints. For example:
+       * - Fruit --> Fruit
+       * - T extends Fruit --> Fruit
+       */
+      const subTypesConstraints = subTypes.map(subType => {
+        const constraint = subType.getConstraint();
+        return constraint === undefined ? subType : constraint;
+      });
+
+      const enumSubTypes = subTypesConstraints.filter(subType =>
+        isEnum(subType),
+      );
+      const baseEnumSubTypes = enumSubTypes.map(subType =>
+        getBaseEnumType(subType),
+      );
+      return new Set(baseEnumSubTypes);
+    }
```

Comment:
Just confirming - this function will filter any non-enum types out, right?
For example if the type is `Fruit | Vegetable | boolean`, then this would return `[Fruit, Vegetable]`?

Would this mean that your rule will error on a comparison like
```ts
enum Fruit { Apple }
declare val: Fruit | boolean;
val === true
```

I coudln't see this covered in the tests.

Reply:
i added that test and it indeed fails. Any suggestions on how to fix? This is getting complicated fast ><

Reply:
this is a case where things get really tricky for sure.
unfortunately there's no "type relationship" API we can use - so there's no way to say like
"is `true` assignable to one of the non-enum types? if yes - ignore it, else report".

So in this case we should probably just bail out and ignore the case for now? We can document it in the readme as a limitation we can (maybe) address later if TS improves its API.

Reply:
Ok, the boolean test passes now

Reply:
oh, you updated your comment, so i'll unresolve this

Reply:
Is `ts.TypeFlags.ESSymbolLike` equivalent to a Symbol? e.g. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol

Reply:
Also, it doesn't seem that there is a `TypeFlag` for a function. So I'll try this:

```ts
/** Equal to: `boolean | object | symbol | function` */
const IMPOSSIBLE_ENUM_TYPES = ts.TypeFlags.BooleanLike | ts.TypeFlags.NonPrimitive | ts.TypeFlags.ESSymbolLike;
```

Reply:
> if the union type contains string, then we must ignore all values of type string. Same goes for number.

Is that true? If I stick the following into VSCode:

```ts
enum Fruit {
  Apple = 0
}
declare const fruitOrNumber: Fruit | number;
```
And then I mouse over `fruitOrNumber`, it reports that the type is number, because the TypeScript compiler resolves the Fruit part away.

I observe the same behavior with the following:

```ts
enum Vegetable {
  Lettuce = "lettuce"
}
declare const vegetableOrString: Vegetable | string;
```

It resolves to just "Fruit".
I also confirmed this inside of an ESLint rule, that calling `leftType.isUnion()` reports false.

Reply:
ok, i managed to make all of those tests pass, im not sure if i did it the way you want though